pkg/activator/handler/handler.go

```diff
@@ -20,33 +20,65 @@ import (
 	"net/url"
 	"strconv"
 	"time"
-
 	"github.com/knative/serving/pkg/activator"
 	"github.com/knative/serving/pkg/activator/util"
 	pkghttp "github.com/knative/serving/pkg/http"
 	"go.uber.org/zap"
 )
 
+const overloadMessage = "throttler overload"
+
 // ActivationHandler will wait for an active endpoint for a revision
 // to be available before proxing the request
 type ActivationHandler struct {
-	Activator activator.Activator
-	Logger    *zap.SugaredLogger
-	Transport http.RoundTripper
-	Reporter  activator.StatsReporter
+	Activator      activator.Activator
+	Logger         *zap.SugaredLogger
+	Transport      http.RoundTripper
+	Reporter       activator.StatsReporter
+	Throttler      *activator.Throttler
+	GetEndpoints   func(activator.RevisionID) (int32, error)
+	UpdateCapacity func(activator.RevisionID, int32) error
 }
 
 func (a *ActivationHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	namespace := pkghttp.LastHeaderValue(r.Header, activator.RevisionHeaderNamespace)
 	name := pkghttp.LastHeaderValue(r.Header, activator.RevisionHeaderName)
-
 	start := time.Now()
+	revID := activator.RevisionID{namespace, name}
+	a.Logger.Debug(revID)
+
+	// ActiveEndpoint() will block until the first endpoint is available.
+	ar := a.Activator.ActiveEndpoint(namespace, name)
+
+	// Potentially update throttler capacity based on the latest state of the endpoints.
+	// This avoid a possible deadlock in case if we missed the updates from the Endpoints informer
+	// and our bookkeeping doesn't reflect the reality.
+	// This could happen because of a restart of the Activator or when a new one is added because of scale out.
+	endpoints, err := a.GetEndpoints(revID)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	err = a.UpdateCapacity(revID, endpoints)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
```

Comment:
Do we really want to do this on each request? Can you elaborate on what kind of deadlock is happening because of the missed update?

Reply:
As discussed via slack - the deadlock could happen when extra instances of activator are added (for example when HPA scales them out), the new instance could loose the updates and the tokens for the breaker in that activator will not get increased. 
As per your suggestion - I will try to move the endpoint update to closer to Breaker creation.